object_store/src/instrumentation.rs

```diff
@@ -0,0 +1,1102 @@
+//! An metric instrumentation wrapper over [`ObjectStoreApi`] implementations.
+
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::{stream::BoxStream, Stream, StreamExt};
+use metric::{Metric, U64Counter, U64Histogram, U64HistogramOptions};
+use pin_project::{pin_project, pinned_drop};
+use time::{SystemProvider, Time, TimeProvider};
+
+use crate::{GetResult, ListResult, ObjectStoreApi};
+
+/// An instrumentation decorator, wrapping an underlying [`ObjectStoreApi`]
+/// implementation and recording bytes transferred and call latency.
+///
+/// # Stream Duration
+///
+/// The [`ObjectStoreApi::get()`] call can return a [`Stream`] which is polled
+/// by the caller and may yield chunks of a file over a series of polls (as
+/// opposed to all of the file data in one go). Because the caller drives the
+/// polling and therefore fetching of data from the object store over the
+/// lifetime of the [`Stream`], the duration of a [`ObjectStoreApi::get()`]
+/// request is measured to be the wall clock difference between the moment the
+/// caller executes the [`ObjectStoreApi::get()`] call, up until the returned
+/// stream is completed (when it yields `Poll::Ready(None)`) or dropped.
+///
+/// # Stream Errors
+///
+/// The [`ObjectStoreApi::get()`] method can return a [`Stream`] of [`Result`]
+/// instances, and returning an error when polled is not necessarily a terminal
+/// state. The metric recorder allows for a caller to observe a transient error
+/// and subsequently go on to complete reading the stream, recording this read
+/// in the "success" histogram.
+///
+/// If a stream is not polled again after observing an error, the operation is
+/// recorded in the "error" histogram.
+///
+/// A stream can return an arbitrary sequence of success and error states before
+/// terminating, with the last observed poll result that yields a [`Result`]
+/// dictating which histogram the operation is recorded in.
+///
+/// # Bytes Transferred
+///
+/// The metric recording bytes transferred accounts for only object data, and
+/// not object metadata (such as that returned by list methods).
+///
+/// The total data transferred will be greater than the metric value due to
+/// metadata queries, read errors, etc. The metric tracks the amount of object
+/// data successfully yielded to the caller.
+///
+/// # Backwards Clocks
+///
+/// If the system clock is observed as moving backwards in time, call durations
+/// are not recorded. The bytes transferred metric is not affected.
+#[derive(Debug)]
+pub struct ObjectStoreMetrics<T, P = SystemProvider> {
+    inner: T,
+    time_provider: P,
+
+    put_success_duration: U64Histogram,
```

Comment:
I would recommend naming all these fields to include the unit. Such as 

```suggestion
    put_success_duration_ms: U64Histogram,
```

A common source of errors I have found is mis interpreting time units and putting the units in the names removes all possible misinterpretations

I see the actual metric names reported have the `_ms` postfix, so adding them to the code would also make things a little more consistent

Reply:
Can't argue with that!